File: libs/sdk-dsl/src/lib/steps/access-steps.ts
import type {
  AccessConfig,
  LinkedAccessConfig,
} from '../interactions/types.js';
import type { JsonObject, JsonValue } from '../core/types.js';
import type { StepsBuilder } from './steps-builder.js';

export class AccessSteps {
  constructor(
    private readonly parent: StepsBuilder,
    private readonly config: AccessConfig,
  ) {}

  requestPermission(): StepsBuilder {
    return this.parent
      .myOs()
      .requestSingleDocPermission(
        this.config.permissionFrom,
        this.config.requestId,
        this.config.targetSessionId,
        {
          read: true,
        },
      );
  }

  revokePermission(): StepsBuilder {
    return this.parent
      .myOs()
      .revokeSingleDocPermission(
        this.config.permissionFrom,
        this.config.requestId,
        this.config.targetSessionId,
      );
  }

  subscribe(): StepsBuilder {
    return this.parent
      .myOs()
      .subscribeToSession(
        this.config.permissionFrom,
        this.config.targetSessionId,
        this.config.subscriptionId,
        'Conversation/Event',
      );
  }

  call(operation: string, request?: JsonValue): StepsBuilder {
    return this.parent
      .myOs()
      .callOperation(
        this.config.permissionFrom,
        this.config.targetSessionId,
        operation,
        request,
      );
  }
}

export class LinkedAccessSteps {
  constructor(
    private readonly parent: StepsBuilder,
    private readonly config: LinkedAccessConfig,
  ) {}

  requestPermission(
    links: Record<string, JsonObject | JsonValue>,
  ): StepsBuilder {
    return this.parent
      .myOs()
      .requestLinkedDocsPermission(
        this.config.permissionFrom,
        this.config.requestId,
        this.config.targetSessionId,
        links,
      );
  }

  revokePermission(): StepsBuilder {
    return this.parent
      .myOs()
      .revokeLinkedDocsPermission(
        this.config.permissionFrom,
        this.config.requestId,
        this.config.targetSessionId,
      );
  }
}

File: references/java-sdk/src/main/java/blue/language/sdk/AccessSteps.java
package blue.language.sdk;

import blue.language.model.Node;
import blue.language.sdk.internal.StepsBuilder;
import blue.language.types.myos.SingleDocumentPermissionGrantRequested;
import blue.language.types.myos.SingleDocumentPermissionRevokeRequested;
import blue.language.types.myos.SubscribeToSessionRequested;

public final class AccessSteps {

    private final StepsBuilder parent;
    private final AccessConfig config;

    public AccessSteps(StepsBuilder parent, AccessConfig config) {
        if (parent == null) {
            throw new IllegalArgumentException("parent is required");
        }
        if (config == null) {
            throw new IllegalArgumentException("config is required");
        }
        this.parent = parent;
        this.config = config;
    }

    public StepsBuilder call(String operation, Object request) {
        return parent.myOs().callOperation(
                config.onBehalfOfChannel(),
                config.targetSessionId(),
                operation,
                request);
    }

    public StepsBuilder callExpr(String operation, String requestExpression) {
        return call(operation, DocBuilder.expr(requestExpression));
    }

    public StepsBuilder requestPermission() {
        return requestPermission("RequestPermission");
    }

    public StepsBuilder requestPermission(String stepName) {
        return parent.emitType(stepName,
                SingleDocumentPermissionGrantRequested.class,
                payload -> {
                    payload.put("onBehalfOf", config.onBehalfOfChannel());
                    payload.put("requestId", config.requestId());
                    payload.put("targetSessionId", config.targetSessionId());
                    payload.putNode("permissions", config.permissionsNode());
                    if (config.subscribeToCreatedSessions()) {
                        payload.put("grantSessionSubscriptionOnResult", true);
                    }
                });
    }

    public StepsBuilder subscribe() {
        return subscribe("Subscribe");
    }

    public StepsBuilder subscribe(String stepName) {
        Node eventList = new Node().items(new java.util.ArrayList<Node>());
        for (Class<?> eventType : config.subscriptionEvents()) {
            if (eventType == null) {
                continue;
            }
            eventList.getItems().add(new Node().type(blue.language.sdk.internal.TypeRef.of(eventType).asTypeNode()));
        }
        return parent.emitType(stepName,
                SubscribeToSessionRequested.class,
                payload -> {
                    payload.put("onBehalfOf", config.onBehalfOfChannel());
                    payload.put("targetSessionId", config.targetSessionId());
                    Node subscription = parent.subscriptionSpec(config.subscriptionId());
                    subscription.properties("events", eventList);
                    payload.putNode("subscription", subscription);
                });
    }

    public StepsBuilder revokePermission() {
        return revokePermission("RevokePermission");
    }

    public StepsBuilder revokePermission(String stepName) {
        return parent.emitType(stepName,
                SingleDocumentPermissionRevokeRequested.class,
                payload -> {
                    payload.put("onBehalfOf", config.onBehalfOfChannel());
                    payload.put("requestId", config.requestId());
                    payload.put("targetSessionId", config.targetSessionId());
                });
    }
}
